Bottom: 46521044b8bc345643a54d631c3a39c1028e14db
Top:    7f5ca6760fb21ad530f3afcc4bf345e03871fec6
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-05-10 17:04:35 +0800

Refresh of circular-buffer

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index c373afb..c843f7b 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1723,9 +1723,10 @@ static int tun_peek_len(struct socket *sock)
 		return 0;
 
 	if (tun->flags & IFF_TX_RING) {
-		unsigned long tail = READ_ONCE(tfile->tail);
-		if (tail != tfile->head)
-			return tfile->tx_descs[tail].len;
+		unsigned long head = smp_load_acquire(&tfile->head);
+		unsigned long tail = tfile->tail;
+		if (head != tail)
+			return tfile->tx_descs[head].len;
 		else
 			return 0;
 	} else {
